### ISSUE TITLE
[idea] Asynchronous validation andThen persist(event) in PersistentActor

### ISSUE THREAD

**Issue by ktoso** [akka/akka#17522]
Inside an PersistentActor, sometimes users may need to `ask` some other actor to perform validation of a message, before it knows if it can persist it or not. Maybe there's a shared "geo-location address checker actor" or something like that.

Currently it is possible to emulate such asynchronous validation by something among the lines of:

```scala
trait ValidationResult { def withHandler(fn: Any => Unit): ValidationResult = this }
final case class Invalid(event: Any) extends ValidationResult
final case class Validated(event: Any, handler: Any => Unit) extends ValidationResult { 
  def def withHandler(handler: Any => Unit) = copy(fn = handler)
}

def receive = {
  case Validated(event, handler) => persist(validated)(handler)
  case Invalid(event) => log.warning("Invalid event! {}, event)
  case event => 
    val handler = (e: Any) => () // some handler anything you like here
    (validator ? PleaseValidate(event))
      .mapTo[Validated]
      .map(_.withHandler(handler)) // set the handler (so we did not have to send it to the validator)
      .pipeTo(self) // send the validation response to self
}
```

However *perhaps* it would be possible to provide a method / trait that would help facilitate this pattern.
Of course, the problem becomes ordering and we would have to think about if we provide two versions (one with stashing, one without) like persist and persistAsync or if we should leave this to users to implement.

This is a loose/fuzzy idea yet comes from a discussion with users who needed this kind of thing.

**Comment by ktoso** [comment]
Related solution - simply recommend putting a validating actor in front of the persistent actor.

**Comment by ktoso** [comment]
Some more notes: 
- implementing what needs to be done with stash/unstashAll is actually not that much code, so perhaps it's enough
- the "delay receiving any more commands" is very simple with out infrastructure in PA
- we would have to allow calling `persist()` in the callbacks though, which currently is not working AFAIR
- adding another method to `PersistentActor` may be a bit much, so consider this as a wild idea.

**Comment by orsonek** [comment]
From Akka 2.4 snapshot documentation: "A persistent actor receives a (non-persistent) command which is first validated if it can be applied to the current state. Here, validation can mean anything, from simple inspection of a command message's fields **up to a conversation with several external services, for example**."

Simple validation calls to external services (actors, ws...) are not possible when persist() can't be invoked from callbacks. 

Workarounds makes code messy (does service need to know Validated / Invalidated messages? if no - should I build another actor which performs validation only and responds with ValidationResult?) or breaks eventsourcing pattern (another validating actor in front of eventsourced one makes the latter just command sourcing IMHO).

**Comment by ktoso** [comment]
On one hand I agree that while we promise in the docs "you can do async validation with external services" we don't have anything built in, so I'd like to provide it. On the other hand I agree with @patriknw that it would grow persistence's API yet again. 

How about we put it at the end of the queue and it could be an experimental "pattern" in contrib for example?

**Comment by patriknw** [comment]
Perhaps I'm missing something? I don't see that this is any different than other things you implement in actor conversations, and why we would need something special for persistence.

    def receiveCommand: Receive = {
      case c: Cmd =>
        validationService ! c
        context.become {
          case OK => persist(event)...; unstashAll(); unbecome();
          case WRONG => ... unstashAll(); unbecome();
          case _ => stash()
        }
    }

**Comment by ktoso** [comment]
Agree that it's not just persistence which has such situations.
My point was that it's something rather common and maybe could be a `akka.actor.pattern` (same like the backoff isn't really a persistence thing, so we'll want to pull it out of persistence, right?).

Not working on it yet, but feel it would be useful to guide newcomers (reminds me of the 2.5 theme, so maybe revisit then).

**Comment by patriknw** [comment]
yes, might fit in the 2.5 theme (there are plenty of ideas and opinions of how to do conversations/protocols with actors, e.g. [aggregator](http://doc.akka.io/docs/akka/2.4-M2/contrib/aggregator.html))

**Comment by ktoso** [comment]
Bah, now that you mention it - the process algebra probably also plays into this!
Moved to 2.5 to keep as a use-case reminder.

**Comment by ktoso** [comment]
Came up on gitter recently: :point_up: [September 29, 2015 4:30 PM](https://gitter.im/akka/akka?at=560aa09e967c1bad7852ba1c)

**Comment by patriknw** [comment]
Good to know. For a future you would pipe the result to self and do the same as above with become and stash.

**Comment by ktoso** [comment]
If someone wants to pick it up please do! It's somewhat of a pet project for me but have not gotten around to work on it.

**Comment by pawelkaczor** [comment]
My take on this: [CollaborationSupport.scala](https://github.com/pawelkaczor/akka-ddd/blob/CollaborationSupport1/akka-ddd-core/src/main/scala/pl/newicom/dddd/messaging/CollaborationSupport.scala)
Sample usage: [DummyAggregateRoot.scala#L107](https://github.com/pawelkaczor/akka-ddd/blob/CollaborationSupport1/akka-ddd-test/src/test/scala/pl/newicom/dddd/test/dummy/DummyAggregateRoot.scala#L107)
